### DISCUSSION TITLE
Challenges with Location Data(Coordinates) Storage and Retrieval

### DISCUSSION THREAD

**Discussion by PuneetSharma52** [adonisjs/core#4688] [Help]
During development, we encountered difficulties in storing and retrieving location coordinates as attributes. A thorough investigation revealed an absence of suitable attributes within the Lucid Model. To address this, the geojson library was integrated from npm, enabling the creation of a Point attribute. This approach facilitated the execution of raw PostGIS queries on the stored data.

The primary objective is to efficiently retrieve a list of users within a specified radius (10 km) of a given user's coordinates. This information will be retrieved or fetched through API endpoints to display nearby users.

Numerous attempts were made to achieve this functionality using various attribute types, including string and geometry, but without success. Guidance is required on effectively modeling location data within the application, including considerations for database migrations and controller implementation.

### Migration
![image](https://github.com/user-attachments/assets/f5273547-19ff-44a9-8fcf-47c5cab3d04c)

### Model
![image](https://github.com/user-attachments/assets/d4f002a2-7b25-4aca-9b27-7de8ae5d18c8)

### Controller
![image](https://github.com/user-attachments/assets/35825002-9216-4c0a-a79a-0e29aa6455ab)

### And also in addition when we are manually seeding the data in the pgsql we are unable to seed it in the location column as showing below for reference.
![image](https://github.com/user-attachments/assets/109c041a-6ec6-4835-b41f-c2053b291c33)

**Comment by tavaresgerson** [comment]
Hi bro,

I'm participating in a project based on coordinates and other proximity calculations between points and polygons and my solution was:

1. Leave the database only to store the data (coordinates) instead of performing the calculations from the database
2. Any calculations were performed in the application layer with [`@turf`](https://turfjs.org/index.html).
3. And I also divided the execution of the Adonisjs application into a cluster (PM2) and it was the best choice

The performance became excellent when the database reached 11.2 GB and the client stopped complaining :-D

**Reply by PuneetSharma52** [reply]
That's great,
In my case i have used raw SQL postgis query that was enough to convert and store the coordinates in the WKT form it was working fine , I also mitigate the SQL injection in that but outpassing the arguments, but in two update queries it was not working,  I will attach that query here, It will be helpful if you can go throw it and provide me solution to mitigate SQL injection in that query 🫡

### Raw SQL query

 await db.rawQuery(
        `UPDATE locations SET location = ST_GeogFromText('SRID=4326;POINT(${longitude} ${latitude})')::geography WHERE customer_id = ?`,
        [customerId]
      )

**Reply by tavaresgerson** [reply]
I think this solves it:

```ts
const sql = "UPDATE locations SET location = ST_GeogFromText('SRID=4326;POINT(?, ?)')::geography WHERE customer_id = ?"
await db.rawQuery(sql, [longitude, latitude, customerId])
```

**Reply by PuneetSharma52** [reply]
I tried this method but it was not working.🙃

**Reply by tavaresgerson** [reply]
Parsing the float values ​​would prevent arbitrary SQL injection:

```ts
const sql = `UPDATE locations SET location = ST_GeogFromText('SRID=4326;POINT(${Number.parseInt(longitude)} ${Number.parseInt(latitude)})')::geography WHERE customer_id = ${Number.parseInt(customerId)}`

await db.rawQuery(sql)
```

Check if it is in accordance with the document: https://postgis.net/docs/ST_GeogFromText.html

Those tubes between the coordinates can't be understood, however just add :smile:

**Reply by PuneetSharma52** [reply]
Ya i have used this thing:- 

```typescript
const [latitude, longitude] = locationString
      .split(' ')
      .map((coord: string) => Number.parseFloat(coord))
    const location = await Location.create({
      id: uuid(),
      customerId,
      type,
    })
    await location.save()
    try {
      await db.rawQuery(
        `UPDATE locations SET location = ST_GeogFromText('SRID=4326;POINT(${longitude} ${latitude})')::geography WHERE customer_id = ?`,
        [customerId]
      )
      return response.status(200).send({ message: 'Location updated successfully' })
    } catch (error) {
      return response.status(500).send({ message: 'Error updating location' })
    }
``` 

i think that's the reason it was showing the errors when i tried to outpass it like -> (?, ?) 🤔
